### DISCUSSION TITLE
Using next/router outside of components

### DISCUSSION THREAD

**Discussion by ezhikov** [vercel/next.js#17046] [Help] [answered]
Documentation states that I have to either use `useRouter` or `withRouter` to access router and I can't find a way to use router outside of components. Thing is, that almost all side effects are handled inside state management library and creating additional state and  `useEffect` just for redirecting user to another page (in my case - error page) seem unnecessary. Is there a way to use router outside of components? I tried to `import {Router} from 'next/router'`, but got `TypeError: Router.push is not a function`.

**Comment by jamesmosier** [comment, marked as answer]
You should be able to import the Router instance and then call `Router.push` on it via:

```js
import Router from 'next/router'
```

**Reply by ezhikov** [reply]
Thank you. My import was wrong.

**Reply by omerman** [reply]
@jamesmosier I must say this is very confusing.. I think it should be changed...
having the 'next/router' file expose variable named `Router` makes code editors suggest along side the default export, which can be very confusing if you don't know about it.

**Reply by ryanflowers** [reply]
+1 on the way these imports work, that really should be fixed as its completely confusing. You have two options

// The router instance
import * as Router from 'next/router';

// Some other objects with only a subset of functionality
import { Router } from 'next/router';

**Reply by sunniejai** [reply]
Are there any downsides of using this instead of using the `useRouter()` hook?

**Reply by JJPell** [reply]
Is there a way to do this for the nextjs app router (next/navigation)?

**Reply by aryankarim** [reply]
I get this.

```
Error: No router instance found.
You should only use "next/router" on the client side of your app.
```

**Comment by vinit-churi** [comment]
> Is there a way to do this for the nextjs app router (next/navigation)?

will (next/router) not work in app router?

**Reply by BrightKingsley** [reply]
No, can't find the push, replace,... methods on it.
Any way to implement this in app router?

**Reply by Tweedle2Dum** [reply]
This seems weird and should not be happening, here is my code for using it outside a functional component (I am using app router). I just imported the Router object from next/router and directly used it.
``` 
import axios from "axios";
import { storage } from "./storage";
import { BASE_URL } from "./constants";
import Router from "next/router";

const API = axios.create({
  baseURL: BASE_URL, 
  timeout: 10000, 
  headers: {
    "Content-Type": "application/json",
  },
});

API.interceptors.request.use(
  (config) => {
    const token = storage.getToken();
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

API.interceptors.response.use(
  (response) => {
    return response;
  },
  (error) => {
    if (error.response.status === 401 && typeof window !== undefined) {
      Router.push('/sign-in')
    }
  }
);
export { API };
```

**Reply by isBatak** [reply]
There is `window.next.router.push()` but it's only available for debugging purposes.
https://github.com/vercel/next.js/blob/74943b67c712271a0254e071282733adccb7cffe/packages/next/src/client/components/app-router.tsx#L417

**Comment by RomanSkrypnik** [comment]
Seems like we will wait for eternity untill Next releases redirect outside of components for app router
